### PR TITLE
Add TRANS_SPRING to Tween

### PR DESCRIPTION
**Pull request by rakkarage**
once again:
https://github.com/godotengine/godot/pull/58999
sorry, thanks

_Bugsquad edited:_ https://github.com/godotengine/godot-proposals/issues/4212

### CODE REVIEW

**Review by TokageItLab — CHANGES_REQUESTED** [review]
I did a quick look into spring tween, but could not find any details about its origin. What I did find is that the formula is similar to the Elastic formula when the spring formula is modified, and that unlike Elastic, the oscillation frequency is accelerated using the cubic of time.

I guessed that the approximate constant was determined to give a good number of about 3 bounds (there may be some additional bounds that are too small to be noticeable) with neat motion, but I don't know what the reason is.

For now, I only review the formulas and leave it to the production team to decide what to do about the license since this algorithm is probably a quote from somewhere.

File: scene/animation/easing_equations.h
Comment (rakkarage) [reply scene/animation/easing_equations.h]: thanks

File: scene/animation/easing_equations.h
Comment (rakkarage) [reply scene/animation/easing_equations.h]: thanks. there are other places in the file that could be reduced in the same way (t * t in one place) (and PI / 2 could be replaced with HALF_PI and PI * 2 could be replaced with TAU

i would prefer to add the "standard" spring tween, not an optimized version that looks 'different'

but i appreciate your time and input

Comment (rakkarage) [reply scene/animation/easing_equations.h]: sorry some of those constants do not exist i guess like HALF_PI

Comment (TokageItLab) [reply scene/animation/easing_equations.h]: It is not necessary to make `pow(s, 2.2f)` to` s * s`, but `s = 1.0f - t` should be implemented. It reduces the computational complexity without changing the result.

Comment (rakkarage) [reply scene/animation/easing_equations.h]: thanks. don't mean to argue... but there are many other situations like that in this file?

back_out_in: 2x c / 2 in one line
bounce_out_in: 2x c / 2 in one line
circ_out_in: 2x c / 2 in one line
cubic: 2x c / 2 in one line
elastic: 2x c / 2 in one line
quart: 2x c / 2 in one line
quad: 2x c / 2 in one line
expo: 2x c / 2 in one line

3x 2 * Math_PI in file
2x Math_PI / 2 in file

the forumula is easier to recognize in its original state
balance optimization with readability
maybe it is fine?
forced premature optimization?
this code is different? or held to different standard then rest of file?

Comment (TokageItLab) [reply scene/animation/easing_equations.h]: Early return and if statement branches do not calculate `c / 2` twice because only the branch is processed.

Comment (rakkarage) [reply scene/animation/easing_equations.h]: thanks

	return in(t * 2 - d, b + c / 2, c / 2, d);

they are all like this... not a branch or a early return but 2x c / 2 on one line? how that not calculate twice? maybe it does a c / 2 on another branch or return but also two on one line which is the same as or worse as having 1 - t twice?

i cannot understand the what u are telling me to do to with `in()`? elastic is the only function that uses `s` and `p` in `in`?

	static real_t in(real_t t, real_t b, real_t c, real_t d) {
		if (t == 0) {
			return b;
		}

		t /= d;
		if (t == 1) {
			return b + c;
		}

		t -= 1;
		float p = d * 0.3f;
		float a = c * pow(2, 10 * t);
		float s = p / 4;

		return -(a * sin((t * d - s) * (2 * Math_PI) / p)) + b;
	}

Comment (TokageItLab) [reply scene/animation/easing_equations.h]: Okay I sent a PR https://github.com/godotengine/godot/pull/77141 which optimizes them.

All I ask of you is simply this change.

```cpp
	t = (sin(t * Math_PI * (0.2f + 2.5f * t * t * t)) * pow(1.0f - t, 2.2f) + t) * (1.0f + (1.2f * (1.0f - t)));
```
to
```cpp
	float s = 1.0f - t;
	t = (sin(t * Math_PI * (0.2f + 2.5f * t * t * t)) * pow(s, 2.2f) + t) * (1.0f + (1.2f * s));
```

If you do it and squash PR correctly then I can personally Approve this.

**Review by TokageItLab — CHANGES_REQUESTED** [review]

**Comment by rakkarage** [comment]
- uses real instead of float like rest of file

File: scene/animation/easing_equations.h
Comment (TokageItLab) [inline scene/animation/easing_equations.h:408]:
```suggestion
	real_t s = 1.0 - t;
```

**Review by TokageItLab — CHANGES_REQUESTED** [review]

File: scene/animation/easing_equations.h
Comment (rakkarage) [reply scene/animation/easing_equations.h:408]: are you sure i should change that one float to a real?
i was thinking about changing the two reals to floats because tho all parameters are real all variables in file are float.
so this is inconsistent?

Comment (TokageItLab) [reply scene/animation/easing_equations.h]: If you want to use a float here, you should write it to `1.0f - t`. Well, since Godot uses double values for delta values for process and animation elapsed time, so it would basically be preferable to have it be `real_t`. There are inconsistencies in some easing, so maybe that could be revised later as a separate PR.

**Review by TokageItLab — APPROVED** [review]
I think now the behavior and code style are fine.

Well, it is unclear as to the original source of this code. I leave it to the production team to decide if there is a licensing concern.

**Review by KoBeWi — APPROVED** [review]
Tween cheat sheet with the new transition:
image
We'll need to update the image in the documentation.

**Comment by kleonc** [comment]
@KoBeWi How is this cheat sheet generated? It could potentially be improved, like lines could be antialiased and the gaps could be eliminated (there are visible gaps e.g. near zero in elastic).

**Comment by KoBeWi** [comment]
With this: https://github.com/wandomPewlin/godotTweeningCheatSheet
Note that the tool does not support Godot 4 yet, I was going send the updated code once this PR is merged.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
